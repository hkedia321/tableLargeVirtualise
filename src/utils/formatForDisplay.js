const numberFormatOptions = {
  minimumFractionDigits: 0,
  maximumFractionDigits: 2,
}

const formatNumberRank = (value) => {
  if (value === 1) return "<span>1<sup>st</sup></span>"
  if (value === 2) return "<span>2<sup>nd</sup></span>"
  if (value === 3) return "<span>3<sup>rd</sup></span>"
  return `<span>${value}<sup>th</sup></span>`
}

export const formatCurrencyForDisplay = (number) => {
  return `&euro;&nbsp;${Number(number).toLocaleString(
    "en",
    numberFormatOptions
  )}`
}

export const formatDateForDisplay = (value) => {
  if (value) {
    const date = new Date(value)
    const month = date.toDateString().substring(4, 7)
    return `${formatNumberRank(date.getDate())}${month} ${date.getFullYear()}`
  }
  return ""
}
