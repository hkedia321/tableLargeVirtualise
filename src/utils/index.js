const numberFormatOptions = { 
    minimumFractionDigits: 0,
    maximumFractionDigits: 2 
};

export const formatNumberForDisplay = (number) => {
return Number(number).toLocaleString('en', numberFormatOptions);
}