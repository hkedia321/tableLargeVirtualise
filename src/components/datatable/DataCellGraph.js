import React from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import sha256 from "crypto-js/sha256"
import Graph from "components/graph"
import * as utils from "utils"

const TotalHeading = styled.h4`
  position: absolute;
  z-index: 1;
  font-weight: normal;
  right: 20px;
  top: -5px;
  text-align: right;
  color: #3f3f3f;
`

const DataCellStyled = styled.div`
  position: relative;
  padding: 10px;
  width: auto;
  padding-top: 30px;
`

const DataCellGraph = React.memo(
  ({ columnObject, rowData, isOpen, style }) => {
    const graphData = rowData.trend.map((dayData) => [
      dayData.day,
      dayData[columnObject.key],
    ])
    const total = rowData.trend.reduce(
      (totalCollect, dayData) => totalCollect + dayData[columnObject.key],
      0
    )
    // eslint-disable-next-line no-console
    console.log("Rendering DataCellGraph")
    return (
      <DataCellStyled
        className="dataCellGraph"
        style={style}
        data-testid="dataCellGraph"
      >
        {isOpen && (
          <>
            <TotalHeading
              dangerouslySetInnerHTML={{
                __html: utils.formatCurrencyForDisplay(total),
              }}
            />
            <Graph data={graphData} total={total} />
          </>
        )}
      </DataCellStyled>
    )
  },
  (prevProps, nextProps) => {
    // if same return true to skip re-rendering
    const h1 = sha256(JSON.stringify(prevProps)).toString()
    const h2 = sha256(JSON.stringify(nextProps)).toString()
    if (h1 === h2) return true
    return false
  }
)

export default DataCellGraph

DataCellGraph.propTypes = {
  columnObject: PropTypes.object.isRequired,
  rowData: PropTypes.object.isRequired,
  isOpen: PropTypes.bool.isRequired,
  style: PropTypes.object.isRequired,
}
