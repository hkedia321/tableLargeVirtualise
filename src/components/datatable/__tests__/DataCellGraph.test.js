import React from 'react';
import { render, cleanup, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import * as utils from 'utils';
import DataCellGraph from '../DataCellGraph';

jest.mock('highcharts-react-official', () => {
    return function DummyHighcharts(
      {options}
    ) {
      return (
        <div id="MockedHighcharts">
          {JSON.stringify(options)}
        </div>
      );
    };
  });

afterEach(cleanup);

const dataCellGraphProps = {
    columnObject: {heading: "App and Installs", key: "aj_app_and_installs"}, 
    rowData: {
        groups: {
            Campaign: {
                metadata: {
                    name: "99733871-f7c5-4b52-9773-1ca69816fc03"
                }
            }
        },
        trend: [
            {
                aj_app_and_installs: 178,
                aj_app_ios_installs: 0,
                aj_coh_0w_and_real_acquisition: 14,
                aj_coh_0w_ios_real_acquisition: 0,
                day: "2020-06-01"
            },
            {
                aj_app_and_installs: 166,
                aj_app_ios_installs: 0,
                aj_coh_0w_and_real_acquisition: 21,
                aj_coh_0w_ios_real_acquisition: 0,
                day: "2020-06-02"
            }
        ]
    },
    isOpen:true
}

it('renders dataCellGraph correctly', () => {
    
    const tableRow = document.createElement('tr');
    const { getByTestId } = render(<DataCellGraph {...dataCellGraphProps} isOpen = {false} />, {
        container: document.body.appendChild(tableRow),
    });
    expect(getByTestId('dataCellGraph'));
    // shouln't have text on close state
    const total = dataCellGraphProps.rowData.trend.reduce((total, dayData) => total + dayData[dataCellGraphProps.columnObject.key], 0)
    const totalAmount = utils.formatCurrencyForDisplay(total);
    expect(getByTestId('dataCellGraph')).not.toHaveTextContent(totalAmount);
})

it('matches snapshot', () => {
    const tree = renderer.create(<DataCellGraph {...dataCellGraphProps} />).toJSON();
    expect(tree).toMatchSnapshot();
})