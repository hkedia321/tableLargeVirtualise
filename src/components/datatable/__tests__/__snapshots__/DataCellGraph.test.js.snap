// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
.c1 {
  position: absolute;
  z-index: 1;
  font-weight: normal;
  right: 20px;
  top: -5px;
  text-align: right;
  color: #3f3f3f;
}

.c0 {
  position: relative;
  padding: 10px;
  width: auto;
  padding-top: 30px;
}

<div
  className="c0 dataCellGraph"
  data-testid="dataCellGraph"
>
  <h4
    className="c1"
    dangerouslySetInnerHTML={
      Object {
        "__html": "&euro;&nbsp;344",
      }
    }
  />
  <div
    id="MockedHighcharts"
  >
    {"chart":{"type":"area","height":"43.75%","margin":0},"credits":{"enabled":false},"title":false,"legend":{"enabled":false},"xAxis":{"labels":{"enabled":false},"title":false,"tickLength":0,"lineColor":"transparent"},"yAxis":{"labels":{"enabled":false},"title":false,"gridLineWidth":0},"plotOptions":{"series":{"fillColor":"#eaebfe","color":"#c8c9ff","marker":{"radius":2.5}}},"tooltip":{},"series":[{"data":[["2020-06-01",178],["2020-06-02",166]]}]}
  </div>
</div>
`;
